### ISSUE TITLE
Integrate AccessKit

### ISSUE THREAD

**Issue by DataTriny** [david-vanderson/dvui#151]
Initially proposed in #121 

Users with various disabilities interact with their computer in very different ways, relying on specific input methods (voice recognition, eye tracking, Braille keyboard...) and/or specific output methods (speech synthesizer, Braille display, screen magnifier...).

To support these use cases, all major operating systems came up with the idea of the accessibility tree. This tree is a semantic representation of the content of all user interfaces currently available on the system. The desktop is the root element of the tree, each top-level window is a branch. Widget containers such as panels are also nodes in the tree. Simple widgets such as buttons or labels are leaf nodes. Each node has a role which indicates what it is (button, label, checkbox, panel, window...) as well as properties and states (name, description, enabled state, checked state...). Good behaving applications are expected to expose their accessibility tree to the operating system, which then exposes the whole tree to assistive technologies.

Assistive technologies are any software which query and manipulate the accessibility tree. Examples include: screen readers, screen magnifiers, voice recognition... Automation testing tools also use the accessibility tree to ensure the presence of UI elements and to perform actions on the application being tested.

If the concept of an accessibility tree is common to all platforms, its technical implementation is very different, making it challenging for UI toolkit developers to support some (if not all) of them.

[AccessKit](https://accesskit.dev) is an answer to this problem. It gives UI toolkit developers a cross-platform way of describing their accessibility tree, hiding away the implementation details of each platform. Because it retains the whole tree, it is a good fit for immediate mode rendering UI toolkits like dvui. Its [C bindings](https://github.com/AccessKit/accesskit-c) make it possible to integrate it into projects written in other languages than Rust.

**Comment by david-vanderson** [comment]
Thank you so much for this detailed overview.  It answers a lot of questions I had about the scope and intention.  I really appreciate it!

**Comment by MichaelBelousov** [comment]
@DataTriny weird question, is it reasonable at all for AccessKit to have a shim over the browser DOM accessibility APIs to work with DVUI in a web environment?

**Comment by MichaelBelousov** [comment]
> What we are trying in AccessKit is to create a hidden DOM, using divs and ARIA attributes. You can find more details in [this discussion](https://github.com/AccessKit/accesskit/discussions/514).

That's actually kind of what I was hoping for/meant. I'll check it out, thanks!

**Comment by phatchman** [comment]
I've started a small POC integrating AccessKit using the C bindings into the SDL backend on Windows to try and understand what changes would be required.

I preface the below with the fact that my total combined lifetime rust experience is now 1 full afternoon.

Everything is compiling OK, but I'm having issues with linking to the rust panic handler? The accesskit examples build fine, so it must be something to do with how I'm linking with zig. I built the accesskit-c libraries from source.

```
error: lld-link: undefined symbol: const type_info::`vftable'
    note: referenced by accesskit.lib(std-6a8c6ea920012cbe.std.ae4a37aaa2130411-cgu.0.rcgu.o):(__rustc::__rust_panic_type_info)
    note: referenced by accesskit.lib(panic_unwind-a8e2f17015f18c0c.panic_unwind.7078d6e88977f384-cgu.0.rcgu.o):(panic_unwind::imp::TYPE_DESCRIPTOR::hd798837bde46c36b)
```

Isn't type_info part of C++ lib, not C? I'm not sure why I have a vftable for type_info in the C bindings?

My guess is that it's because I could only get the "msvc" target for aarch64 Windows working with rust, but I can only get dvui to build for a "gnu" target. See below for details on this.

When I try and build the accesskit-c library with target=aarch64-pc-windows-gnullvm, I get the error below. There's no plain aarch64-windows-gnu target like there is for x86_64, unfortunately:
```
   Compiling accesskit-c v0.17.0 (C:\src\accesskit-c)
error: linker `aarch64-w64-mingw32-clang` not found
  |
  = note: program not found
```
I've got MSYS2 and clang installed, but I'm not sure what rust is looking for here?

Edit: I think passing /GR- to turn off RTTI may help here? But I can;t for the life of me work out how to pass that flag to the msvc compiler when building accesskit-c with cargo? I've tried adding the flag to CMakeLists.txt and various other promising looking places. But no success so far.

## Findings so far
- I've put together something very simple that just adds the main window and a title to the node tree. I haven't seen any real blockers, but early days. 
- DVUI can provide stable node-id's between frames, which seems to be key to making the integration work.
- Not totally sure yet the best way to integrate the library itself:
    - The library is obviously in rust and we don't really want to introduce rust / cargo into the DVUI build process.
    - I think that ideally we want to link to the static lib, but to do that we would need to supply our own pre-built accesskit static lib targets for all common dvui build targets. (AccessKit has static libs hosted on github, but not for all the targets we'd want).
    - It looks like rust mostly covers the targets we'd normally support with DVUI. (outside of limited windows aarch64 targets).
    - If we went the DLL route instead, we'd have to check for the DLL and disable accessibility if it's not present. But I feel even adding this one step of requiring people to bundle a DLL makes it less likely that people will ship with accessibility enabled. 
- Brings in 2 additional dependencies on windows ws2_32.dll and Userenv.dll
- The dvui build.zig seems broken when you pass the 3rd component of the target triple, e.g. -gnu, -musl, -msvc. At least on my machine.
    -  I _think_ the issue is that it makes the SDL build think it is being cross-compiled and thus won't search the standard OS include/lib paths. i.e. passing -Dtarget=aarch64-windows works, but aarch64-windows-gnu etc doesn't.
- There are attributes that users would typically need to set for accessibility which we don't capture today (e.g. alt text for an image / icon).
- There are cases where widgets need to be associated with each other that aren't parent/children of each other e.g. a label describing a text entry widget.
- Lots of other things I've yet to think about, such as widgets that only exist in some frames (i.e. within expanders / accordions)
    - But even then, they will keep the same id when they are re-created, so maybe isn't an issue.

Edit 1:
- For SDL (and probably others), we need to create the window hidden, then initialize accesskit, then show the window. Minimal changes.
- Default panic handler is to exit on OOM / Error. This is different to our approach in DVUI, which is to only panic as a last resort and try to do something sensible until next frame otherwise. But we still do have plenty of `@panics`, so I doubt this is realistically an issue for anyone.

**Comment by phatchman** [comment]
hmm. What do we think of the chromium license that comes with accesskit?

See the binary distribution clause. Though, it is not clear to me what is counted as binary distribution? Just the compiled libraries, or any application which includes the compiled libraries? i.e. does this become a viral license, extended to DVUI user's applications if they link in accesskit?

@DataTriny Are you able to clarify that at all?

```
// Copyright 2015 The Chromium Authors. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//    * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//    * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//    * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```

**Comment by DataTriny** [comment]
Hi @phatchman,

Regarding licensing: I think the dvui project would have to make it clear for its users that it incorporates AccessKit, mentioning the dual MIT OR Apache-2.0 license and the BSD-3-Clause.

Users who distribute binaries whould then have to include license text from AccessKit somewhere in their app like they would for dvui itself. They can choose to use AccessKit under the terms of either MIT OR Apache-2.0 so only one of these would have to be included, but the BSD-3-Clause notice would have to be there regardless. AccessKit licenses don't prevent users from choosing any license they like for their product, but copyright notice must be found somewhere, and this also apply to dvui's MIT.

**Comment by david-vanderson** [comment]
@phatchman Thanks very much for the detailed findings!

Thanks @DataTriny and I agree - not viral, but we do need to take care to preserve the copyright and license text.

**Comment by DataTriny** [comment]
Also, if dvui needs pre-built binaries for platforms and architectures that our CD pipeline don't cover currently, please let me know. If it doesn't require substancial changes then we can build these for you, assuming using AccessKit through the C bindings is the right choice for dvui of course. I quickly tried building for Windows ARM64 with the GNU toolchain but did not succeed so far.

**Comment by david-vanderson** [comment]
> Also, if dvui needs pre-built binaries for platforms and architectures that our CD pipeline don't cover currently, please let me know. If it doesn't require substancial changes then we can build these for you, assuming using AccessKit through the C bindings is the right choice for dvui of course. I quickly tried building for Windows ARM64 with the GNU toolchain but did not succeed so far.

Thank you will do!

**Comment by DataTriny** [comment]
I think we'll have to wait for the next release of Corrosion to get Windows ARM64 GNU support through our CMake build process. However I was able to successfully build static and dynamic libs on Ubuntu targetting Windows ARM64 by using our Meson build process.

**Comment by phatchman** [comment]
Thanks very much for your replies. 

Using the msvc release build seems to link OK for me. Looks to have just been an issue with additional checks included in debug builds. I can start making progress again.

And I can confirm the panic handler is actually working! Nice error message.

```
thread '<unnamed>' panicked at C:\Users\phatc\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\accesskit_windows-0.29.0\src\subclass.rs:168:13:
The AccessKit Windows subclassing adapter must be created before the window is shown (made visible) for the first time.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
sdl3-standalone
```
Edit: OK, minimal POC works. I'll look at getting the Basic Widgets examples page accessible and we can go from there.

<img width="892" height="568" alt="Image" src="https://github.com/user-attachments/assets/18d67367-3dd1-4815-aa1d-a4cecd42cb15" />

Just to give an idea, file sizes for the demo with and without accesskit linked are below. So ~2.2MB. Seems reasonable. Maybe we can give an option to disable if that would cause an issue for people.
```
without - 13200384
with    - 13427200
```
Edit: That size includes the symbols for stack traces. I'll have do a comparison of stripped versions as it is probably much less.

**Comment by phatchman** [comment]
@DataTriny Thanks for getting back to me so quickly. 

Could you check if I'm understanding correctly how we might use the API? 
Is there another non-rust IM UI that has already integrated with access kit, preferable written in C. While I'd love to learn rust, it's a pretty big commitment!

I'm ignoring actions for now. I'll just concentrate on getting the node tree populated correctly first.

My rough plan is as follows:
1) Before the SDL window is shows, call `accesskit_windows_subclassing_adapter_new` with the window handle and callbacks.
2) The first tree from the `activation_handler` callback will just contain the main window as no widgets would have been created yet.
3) We'll then show the window.
4) On the second and subsequent frames, we'll add child widgets of the main window using `accesskit_node_push_child`. And if those widgets have child widgets, we'll push those as children of the parent etc. 
5) I'm expecting I can always just report the current state for a widget (.e.g. if a button is pressed or checkbox checked) each frame, rather than only on state change. i.e. We won't necessarily know if someone changes some label text between frames. While we would know if a button state changed, it would be cleaner if we could just report current button state each frame.
6) Will need to deal with new and removed widgets here somehow....
7) After pushing the child widget, we'll set the properties for that widget e.g. `accesskit_node_set_bounds`, `accesskit_node_add_action` etc.
8) Once that is done, we'll build the tree update using `accesskit_tree_update_with_capacity_and_focus`, `accesskit_tree_new`, `accesskit_tree_update_set_tree`
9) The child nodes then need to be added to the tree using `accesskit_tree_update_push_node`

Or is there some way we can just build a new tree each time, preserving the same node IDs, rather than doing updates? That would just save working out which widgets were removed and/or added since the previous frame.

**Comment by DataTriny** [comment]
@phatchman AFAIK the only public users of AccessKit through the C bindings are GTK and Godot, none are immediate mode.

2. If you can only push the window node before showing the SDL window, you might prefer to return `NULL` from the activation handler and we'll generate a placeholder tree for you. Just remember to push a full tree the first time you call `accesskit_windows_subclassing_adapter_update_if_active`.
3. ...
4. ...
5. Your understanding is correct: you can report the current state of a checkbox for instance on every frame and we will figure out the difference ourself.
6. When you remove and/or add nodes, remember to also include their parent in the tree update with its children property reflecting that.

> Or is there some way we can just build a new tree each time, preserving the same node IDs, rather than doing updates? That would just save working out which widgets were removed and/or added since the previous frame.

I'm not sure I understand what you want here. If you prefer pushing the entire tree each time, that is fine, as long as IDs are stable across updates. If you prefer to not push nodes that haven't changed this is also fine, and slightly faster of course. Just remember to push the parent of added/removed nodes.

**Comment by phatchman** [comment]
Hi @DataTriny Do you know why the SDL example wraps most of the calls to the C API in a mutex e.g.:
```
  window_state_lock(state);
  accesskit_tree_update *update = window_state_build_initial_tree(state);
  window_state_unlock(state);
```
I'm getting some inconsistent behaviour and wondering if it is something to do with me ignoring locking? But it's not clear to me why there would be a multi-threading concern here? 
Are the callbacks for the "update" functions called on a different thread to the SDL main thread?

Edit: The weird behaviour was unrelated. But still curious if the locking is required here?

Thanks,

Paul

**Comment by phatchman** [comment]
One of the other things I've also been wrestling with is mapping of a dvui button onto the tree.

In DVUI a button is really just something clickable. The button contains a label widget and/or an image widget or almost any collection of widgets you like.

From an accessibility point of view is it fine to construct a button node that has child nodes for the labels and images, rather than setting the label / image on the button node itself? What would be a good way to test things like that out myself? Is using a screen reader sufficient?

Thanks.

**Comment by DataTriny** [comment]
> Are the callbacks for the "update" functions called on a different thread to the SDL main thread?

This depends on the platform:

- On macOS all callbacks are called on the UI thread.
- On Linux all callbacks are called on an another thread managed by AccessKit.
- On Windows the activation callback will be called on the thread that owns the window, however there is no guarantee on where the action callback will be called.

So yes, protecting the UI state from concurrent access is required.

**Comment by DataTriny** [comment]
> From an accessibility point of view is it fine to construct a button node that has child nodes for the labels and images, rather than setting the label / image on the button node itself?

Yes. You would typically have to call `accesskit_node_set_labelled_by` on the parent to indicate the relationship. However buttons are a bit special in the sense that their children will automatically be assigned this relationship. If you add a labelled image as child of a button, then this button will be labelled with the label of the image.

> What would be a good way to test things like that out myself? Is using a screen reader sufficient?

To test such behavior, screen readers would be the best. For higher level checks such as checking the overall structure of your tree or if you are familiar enough with how your nodes should be laid out, there are accessibility inspector tools: Accessibility Insights for Windows, Accessibility Inspector on macOS, Accerciser and Accessibility Inspector on Linux.

**Comment by phatchman** [comment]
Things seems to be progressing well. Just continuing to fill out the tree and properties for all the widgets and associated dvui.zig fucntions. Most of the issues are me trying to understand what properties should be populated for each role. 

@david-vanderson The following code is used for buttons, radio buttons and checkboxes:
```
    var bw = ButtonWidget.init(src, .{}, options.strip().override(options));
```
I don't understand the purpose of that line. If the options are stripped and overridden with themselves, doesn't it just end up being the original options? I ask because I want to pass "accessibility options" as an option override to that button to set its role correctly i.e. depending on whether it is acting as a checkbox, radio etc.

**Comment by MichaelBelousov** [comment]
Thank you for this great work so far @phatchman! While 2MB executable size growth isn't *terrible* for wasm, I'm going to add a separate issue to have a lighter weight integration with the browser that probably doesn't use access kit.

**Comment by phatchman** [comment]
Here's a preview of the Basic Widget's pane. Menus and sliders still need some work and the more complex widgets like grid and tree will be a little trickier.

<img width="484" height="223" alt="Image" src="https://github.com/user-attachments/assets/61b918b8-3709-4268-b192-78670b03838d" />

<img width="463" height="198" alt="Image" src="https://github.com/user-attachments/assets/3fecac21-d21a-4fcc-8b49-229cace055bd" />

<img width="514" height="261" alt="Image" src="https://github.com/user-attachments/assets/3333043c-708d-46a0-adaf-c239a61a6948" />

<img width="599" height="969" alt="Image" src="https://github.com/user-attachments/assets/dd32b495-b246-47bc-aef0-ccc1106025fa" />

@david-vanderson One more question. Is there a good way to work out if a widget will be shown? There's no point adding things to the accessibility tree that the user can't see (e.g. scrolled off grid cells).

Is it just a case of getting the borderRectScale and checking if that is inside the window's rect? I don't care if it is invisible behind something, just that it is not scrolled or cropped off the screen.

Actually, that makes me think I should be clipping the physical rect's I'm passing to access kit. Otherwise, a button that is cropped won't be represented correctly.

**Comment by phatchman** [comment]
> Thank you for this great work so far [@phatchman](https://github.com/phatchman)! While 2MB executable size growth isn't _terrible_ for wasm, I'm going to add a separate issue to have a lighter weight integration with the browser that probably doesn't use access kit.

I doubt it will actually be that large in the end. The current libraries have enough debug info in them to do fairly detailed stack traces I believe we can build the library without it and should significantly reduce the size. I believe I saw someone mention ~600K, but let's see what it ends up being in reality.

**Comment by MichaelBelousov** [comment]
> I doubt it will actually be that large in the end. The current libraries have enough debug info in them to do fairly detailed stack traces I believe we can build the library without it and should significantly reduce the size. I believe I saw someone mention ~600K, but let's see what it ends up being in reality.

My bad I think then for assuming you checked the release binary size.

Please comment on the new issue maybe then?
https://github.com/david-vanderson/dvui/issues/593
600K isn't that bad actually, just I'm thinking about the web backend where we'll need to implement a DOM shim anyway in access kit so almost might as well directly do it in DVUI to remove the overhead of access kit.  Although at 600K I might just prefer to implement it in access kit and help them reduce their binary size when compiling for the browser

**Comment by phatchman** [comment]
> Please comment on the new issue maybe then? [#593](https://github.com/david-vanderson/dvui/issues/593) 

I'll try and get some numbers tomorrow and add to your issue. My numbers were all with release builds, but IIRC you have to add some flag to the accesskit release build to exclude debug info from it.

**Comment by DataTriny** [comment]
@MichaelBelousov The numbers you mention only apply on *nix platforms where we embed a D-Bus library. Binaries for Windows and macOS are much smaller, expect the same when we will add support for the web. I know that binary size is a big deal for the browser, this is something I'll watch closely.

**Comment by phatchman** [comment]
@david-vanderson I'd like to get your thoughts on my plan for implementing actions.

Actions are things like clicking on a button, scrolling, setting, selecting and replacing text, changing focus. These actions can arrive on a separate thread to the gui thread. Actions will be for a "node", which maps directly to a dvui widget id.
Each widget can advertise which actions it supports.

I plan to put the actions in a thread-safe queue to be processed in the next frame.
Each widget's processEvents() would search this frame's copy of the "accessibility actions queue". Any actions for the widget's widget id. would be treated in the same way processing the actual event, e.g. a click event.

So, for button widget:
```
pub fn processEvents(self: *ButtonWidget) void {
    self.click = dvui.clicked(self.data(), .{ .hovered = &self.hover });
}
```
Would be changed to be `self.click = dvui.clicked(....) or "has an accessibility click action"` 

Similarly, TextEntryWidget could do the select and replace etc in its processEvents().

Does that sound reasonable to you? Any other ideas?

The only other thing I know I need to work out is the best way to handle changing focus when an action happens on the widget. I assume when a click action happens on a button, we'd also want to change focus to the button, just as if a mouse clicked it.

Thanks.

**Comment by david-vanderson** [comment]
> [@david-vanderson](https://github.com/david-vanderson) The following code is used for buttons, radio buttons and checkboxes:
> 
> ```
>     var bw = ButtonWidget.init(src, .{}, options.strip().override(options));
> ```

Sorry - this is confusing.  The intent is to use `strip()` to zero-out some fields (margin/border/padding/corner_radius/background) so that we don't get ButtonWidget's defaults.  So after `strip()` we know those fields are set and won't get overridden inside ButtonWidget.

Then we override that to replace those zeroed-out fields in the case where the application provided those fields.

(That code was written before we had the `clicked()` helper - all it really wanted was ButtonWidget's event handling.  So we can rewrite it to not use ButtonWidget)

Does that make sense?

**Comment by david-vanderson** [comment]
Awesome progress!

> [@david-vanderson](https://github.com/david-vanderson) One more question. Is there a good way to work out if a widget will be shown? There's no point adding things to the accessibility tree that the user can't see (e.g. scrolled off grid cells).
>
> Is it just a case of getting the borderRectScale and checking if that is inside the window's rect? I don't care if it is invisible behind something, just that it is not scrolled or cropped off the screen.

`WidgetData.visible()` and `Rect.Physical.visible()` are intended for this purpose - they check a rect (usually the border rect) against the current clipping rect.

One wrinkle is that you can tab to a widget that is not visible on that frame, using tab navigation to scroll (if a widget is newly focused on a frame, it calls `scrollTo()`.  I don't know the right answer here:
* only add visible widgets?
* additionally add non-visible widgets that have a tab index?
* additionally add non-visible widgets by expanding the clipping region for this check?

> 
> Actually, that makes me think I should be clipping the physical rect's I'm passing to access kit. Otherwise, a button that is cropped won't be represented correctly.

That makes sense to me, unless we might also add non-visible widgets sometimes.  Then I don't know what is right.

**Comment by david-vanderson** [comment]
> [@david-vanderson](https://github.com/david-vanderson) I'd like to get your thoughts on my plan for implementing actions.
> 
> Would be changed to be `self.click = dvui.clicked(....) or "has an accessibility click action"`
> 
> Does that sound reasonable to you? Any other ideas?

I'm not totally sure, but my gut says we should run this through the existing Event queue.  We can pull them all off the inter-thread queue at the start of the frame?

We did that with our initial testing stuff, and this feels similar.

So basic question: if we rewrite an "accessibility click action" so that dvui sees it as a "move mouse, mouse focus, mouse down, mouse up" sequence, are we losing important information that a widget might use?

> Similarly, TextEntryWidget could do the select and replace etc in its processEvents().

This is a bit harder, because `dvui.Event` doesn't (yet) have events that represent those kind of higher-level logical actions.  Probably it needs them, and not just for this.  I don't currently have an answer to how to have a menu item that does "paste" into the last focused textEntry for example.

So my guess is that we want to enhance the Event loop rather than deal with it separately?

**Comment by phatchman** [comment]
> So my guess is that we want to enhance the Event loop rather than deal with it separately?

I think it is possible, but  I was uncertain about a few things:
1) What mouse position do we use? 
  - widget data doesn't exist at start of frame and all I get from the action is the widget id.
  - But I can get the screen rect for that node from AccessKit, and use the middle of that as the mouse pos.
  - Is clicking in the middle always going to work for all widgets?
2) Entering a value into a widget would be sending a series of key down, key ups for each letter.
  - That could be a lot of events for a long string.
  - So would we need a new "set value" event that has to be handled separately anyway?
3) Impact on drag and drop from additional mouse moves
  - This one is probably ok if the mouse move gets filtered out in eventMatch. I could make sure events from actions are only seen by the intended widget.
4) If the widget already has mouse events
  - It will get mouse events from 2 sources in 1 frame. 
  - In something like a menu, I can see this causing issues, getting 2 click events in 1 frame.

**Comment by david-vanderson** [comment]
> > So my guess is that we want to enhance the Event loop rather than deal with it separately?
> 
> I think it is possible, but I was uncertain about a few things:

I agree - there are a lot of details that are uncertain, but it's still my best bet.  Some quick thoughts:

>     1. What mouse position do we use?
> 
>     * widget data doesn't exist at start of frame and all I get from the action is the widget id.
> 
>     * But I can get the screen rect for that node from AccessKit, and use the middle of that as the mouse pos.
> 
>     * Is clicking in the middle always going to work for all widgets?

That sounds reasonable, so we should try that first.  Can figure out something else if it doesn't work (like saving the screen position ourselves somehow).

>     2. Entering a value into a widget would be sending a series of key down, key ups for each letter.
> 
>     * That could be a lot of events for a long string.
> 
>     * So would we need a new "set value" event that has to be handled separately anyway?

The existing "text" event is a whole string at the same time, so that might work as a "set value" event.  For example, in textEntry almost all the text comes via a "text" event.  But enter/backspace/delete are handled via the "key" events.

We probably need some experimentation here.

> 
>     3. Impact on drag and drop from additional mouse moves
>
>     * This one is probably ok if the mouse move gets filtered out in eventMatch. I could make sure events from actions are only seen by the intended widget.

I'm totally at a loss for how drag-n-drop works in various accessibility situations.  Does anyone know how that works?

> 
>     4. If the widget already has mouse events
> 
>     * It will get mouse events from 2 sources in 1 frame.
> 
>     * In something like a menu, I can see this causing issues, getting 2 click events in 1 frame.

My guess is that events coming from accesskit would generally not happen at the same time as normal events?  But generally it is okay for widgets to get a series of events.  For example in a low framerate scenario you can defintely click on things multiple times in a single frame.  If this causes problems then we need to fix those widgets.

**Comment by phatchman** [comment]
> One wrinkle is that you can tab to a widget that is not visible on that frame, using tab navigation to scroll (if a widget is newly focused on a frame, it calls `scrollTo()`. I don't know the right answer here:

I don't think that's going to be an issue. The widget will just get added to the acessibility tree next frame. I build a new tree every frame and accesskit seems smart enough to work out what nodes need to be added / removed from the previous frame.

And I just did some more playing to confirm we really do need to clip and only add visible.

**Comment by phatchman** [comment]
At least for button clicks, using events is working well. Edit: And now for setting text values as well. 

Edit2: Well actually we need to extend text a bit I think. Since text appends where I also need the ability to "set". So should I just introduce a new flag that is append / replace to the text event?

I will need to discuss with you a better way to do the event creation. I found the existing window.addEventXXX won't work for this. 

e.g. addEventPointer does 
```
    const widget_id = if (self.capture) |cap| cap.id else null;
    const winId = self.subwindows.windowFor(self.mouse_pt);
```
But I need to send events to widgets independent of where the physical mouse cursor is and I already know which widget it is for. For now, I'm creating the events and directly adding them to window.events, which probably misses some important logic from those addEventXXX functions, but I've not seen any bad effects so far.

I'm also using 0 for the "dp" in the motion event figuring I don't want to change any relative mouse move calculation by moving the mouse to some other part of the screen and back. Not sure if that is a good idea or not. Edit: Actually, do I really need the motion event? The press and release have the mouse point in them and I doubt anyone is using motion for click detection? 

I think I'll focus on getting what I've got worked up into a draft PR so we can go through some of those issues. Then it should just be incrementally adding support for more widgets and actions.

Edit3: It is a pretty common requirement to want to label things, e.g. associate a lbel with a text entry box. This is fine I've implemented  .labeled_by and a .label_for in the AccessibilityOptions fields for dvui.Options.

The issue is getting the id of the label widget is hard with the current API. You have to use data_out and I fear that if we make doing accessibility correctly even mildly difficult, then no one is going to do it. But I'm not sure what a good way to fix that would be outside of just returning the widget id, which doesn't seem great to me. 

Maybe we can just have something that gives the last created widget id?

Edit 4 (sorry): Actually,  for the simple case, e.g. labels for text boxes, I've implemented a .label option. It is much easier for someone to just pass in a label string directly rather than using `.labelled_by`. e.g. below: (I'm testing out using the name `.a11y` instead of `.accessibility` as the field name.)
```
        _ = dvui.slider(@src(), .{ .dir = .vertical, .fraction = &slider_val }, .{
            .expand = .vertical,
            .min_size_content = .{ .w = 10 },
            .corner_radius = dvui.Rect.all(100),
            .a11y = .{ .label = "Sliders2" },
        });
```

**Comment by david-vanderson** [comment]
Excellent!  Have read the PR text, and will dig into it after responding here:

> At least for button clicks, using events is working well. Edit: And now for setting text values as well.
> 
> Edit2: Well actually we need to extend text a bit I think. Since text appends where I also need the ability to "set". So should I just introduce a new flag that is append / replace to the text event?

I don't know if it should be a flag on the event, or if we should have a separate "select all" event.  Seems like eventually we'd need the separate event anyway, so I will try to figure that out.

> 
> I will need to discuss with you a better way to do the event creation. I found the existing window.addEventXXX won't work for this.
> 
> e.g. addEventPointer does
> 
> ```
>     const widget_id = if (self.capture) |cap| cap.id else null;
>     const winId = self.subwindows.windowFor(self.mouse_pt);
> ```
> 
> But I need to send events to widgets independent of where the physical mouse cursor is and I already know which widget it is for. For now, I'm creating the events and directly adding them to window.events, which probably misses some important logic from those addEventXXX functions, but I've not seen any bad effects so far.
> 
> I'm also using 0 for the "dp" in the motion event figuring I don't want to change any relative mouse move calculation by moving the mouse to some other part of the screen and back. Not sure if that is a good idea or not. Edit: Actually, do I really need the motion event? The press and release have the mouse point in them and I doubt anyone is using motion for click detection?
> 
> I think I'll focus on getting what I've got worked up into a draft PR so we can go through some of those issues. Then it should just be incrementally adding support for more widgets and actions.

Yep looks like those addEventXXX functions need to grow more options.  For now directly creating the events sounds good so we can figure out what is needed.  I think you are probably right that we can skip the motion events in this case.

We might be able to reuse Event.target_widgetId as a more general purpose "this event is for this widget", which I think should work unless a different widget grabs mouse capture or focus (for key/text events) in the same frame.  That could allow us to send events without sending focus events first.  I don't know if that is a good idea or not - seems like if you are sending an event to a widget, it should probably get focus?

> 
> Edit3: It is a pretty common requirement to want to label things, e.g. associate a lbel with a text entry box. This is fine I've implemented .labeled_by and a .label_for in the AccessibilityOptions fields for dvui.Options.
> 
> The issue is getting the id of the label widget is hard with the current API. You have to use data_out and I fear that if we make doing accessibility correctly even mildly difficult, then no one is going to do it. But I'm not sure what a good way to fix that would be outside of just returning the widget id, which doesn't seem great to me.
> 
> Maybe we can just have something that gives the last created widget id?
> 
> Edit 4 (sorry): Actually, for the simple case, e.g. labels for text boxes, I've implemented a .label option. It is much easier for someone to just pass in a label string directly rather than using `.labelled_by`. e.g. below: (I'm testing out using the name `.a11y` instead of `.accessibility` as the field name.)
> 
> ```
>         _ = dvui.slider(@src(), .{ .dir = .vertical, .fraction = &slider_val }, .{
>             .expand = .vertical,
>             .min_size_content = .{ .w = 10 },
>             .corner_radius = dvui.Rect.all(100),
>             .a11y = .{ .label = "Sliders2" },
>         });
> ```

That's a good point.  Maybe we can put in some kind of heuristic (like "last label seen") that is used if an accessiblity label is not given.  Not perfect but could be better than nothing, and I agree we need to make this as easy as possible.

**Comment by phatchman** [comment]
>That's a good point. Maybe we can put in some kind of heuristic (like "last label seen") that is used if an accessiblity label is not given. Not perfect but could be better than nothing, and I agree we need to make this as easy as possible.

Yeah, can try something like that. The issue I see is that not all labels are 'labels'. E.g. Creating 3 widgets, a label, a button, and text entry. The last label for the text entry would be the button text. But maybe we can do something with the accesskit roles to distinguish between different types of labels.

**Comment by phatchman** [comment]
> We might be able to reuse Event.target_widgetId as a more general purpose "this event is for this widget", which I think should work unless a different widget grabs mouse capture or focus (for key/text events) in the same frame. That could allow us to send events without sending focus events first. I don't know if that is a good idea or not - seems like if you are sending an event to a widget, it should probably get focus?

From testing, it seems that it is normal for the widget that receives the action also to get focus.  We should be raising to front the main window, and any sub-window containing the widget on processing of an action.

We also need to consider that an action can be received on a widget that normally would not be able to process that action. e.g. in the 'Dialogs' demo window I can send an invoke/click to the buttons in that window even when a modal dialog is showing. With the current code, those buttons will process their actions and show another dialog.

**Comment by david-vanderson** [comment]
> > We might be able to reuse Event.target_widgetId as a more general purpose "this event is for this widget", which I think should work unless a different widget grabs mouse capture or focus (for key/text events) in the same frame. That could allow us to send events without sending focus events first. I don't know if that is a good idea or not - seems like if you are sending an event to a widget, it should probably get focus?
> 
> From testing, it seems that it is normal for the widget that receives the action also to get focus. We should be raising to front the main window, and any sub-window containing the widget on processing of an action.

Okay that sounds doable.

> 
> We also need to consider that an action can be received on a widget that normally would not be able to process that action. e.g. in the 'Dialogs' demo window I can send an invoke/click to the buttons in that window even when a modal dialog is showing. With the current code, those buttons will process their actions and show another dialog.

This is likely one of those things that's handled by the extra code inside the addEventXXX functions, so should shake out eventually.  We should check what happens in regular windows applications in that scenario (sending an action to a widget below a modal dialog).

**Comment by DataTriny** [comment]
Given the current state of the AccessKit integration, I don't think there is a need to keep this generic issue open. Remaining gaps or bugs will be better tracked in dedicated issues.

I am very happy I can close it now. Huge thanks to @phatchman who did the vast majority of the work! Accessibility being an ongoing process and not a box that can be ticked once and for all, I'll stick around whenever you folks need support.

**Comment by david-vanderson** [comment]
Thank you so much @phatchman for all the hard work, and @DataTriny for holding our hand through this and for extensive testing and feedback!
